working/macros/feature-specification.md

```diff
@@ -304,20 +304,109 @@ class, allowing you to build up your code fragments however you like.
+refer to them — it may call macro-generated functions, read macro-generated
+fields, construct macro-generated classes, etc. This means that before macros
+are applied, code may contain identifiers that cannot be resolved. This is not
+an error. Any identifier that can't be resolved before the macro is applied is
+allowed to be resolved to a macro-produced identifier after macros are applied.
+
+Macros are not permitted to introduce declarations that directly conflict with
+existing declarations in the same library. These rules are the same as if the
+code were hand written.
+
+Macros may also end up adding new declarations which shadow existing symbols in
+the library, but don't directly conflict. In this case we want to ensure that
+the intent of any user written code is always clear. Consider the following
+example:
+
+```dart
+int get x => 1;
+
+@generateX 
+class Bar {
+  // Generated: int get x => 2;
+
+  // Should this return the top level `x`, or the generated instance getter?
+  int get y => x; 
+}
+```
+
+There are several potential choices to we could make here:
+
+1.  We could say that any identifier that can be resolved before macro
+    application keeps its original resolution (so `x` would still resolve to the
+    original, top level `x`).
+2.  We could re-resolve all identifiers after the macros are applied, which can
+    possibly change what they resolve to (in this case `x` would resolve to the
+    generated instance getter `x`).
+3.  We could make it some kind of error for a macro to introduce an identifier
+    that shadows another.
+4.  We could make it error to use an identifier shadowed by one produced by a
+    macro.
+
+The first two choices could be very confusing to users, some will expect one
+behavior while others expect the other. The third choice would work but might be
+overly restrictive. The final option still avoids the ambiguity, and is a bit
+more permissive than the third.
+
+It similarly also is not allowed for one macro to produce an declaration that
+the identifier resolves to and then another macro to produce another declaration
+that then shadows that one. In other words, any hand-authored identifier may be
+resolved at any point during macro application, but it may only be resolved
+once.
+
+At the same time, a macro application in a library must be able to shadow
+identifiers in other libraries, in order to allow for modular compilation. It
+shouldn't be the case that a library can cause an error in one of the libraries
+that it imports.
+
+These constraints produce this rule:
+
+*   It is a compile-time error if any hand-authored identifier in a library
+    containing a macro application would bind to a different declaration when
+    resolved before and after macro application. In other words, it is a
+    compile-time error if a macro introduces an identifier that shadows a
+    hand-authored identifier that is used in the same library.
```

Comment:
What about:

```dart
@generateXTopLevelVariable
library foo;

@generateXGetter
class C {
  int get y => x;
}
```

Is this an error? If `@generateXGetter` were at some point in the future be able to introspect over the body of `y`, what would it see that `x` has been resolved to, if anything?

I think the answer is probably that this is allowed—multiple macro applications can produce identifiers that *do* shadow each other. The user's expectation is only how identifiers in the actual text file are resolved and how identifiers after all macro applications are done are resolved. In both cases, the behavior is well-defined. It might be worth calling this out explicitly.

In the example, introspection should see it as an unresolved identifier.

Reply:
Ya I have the same reasoning for calling out that its only hand-authored things that can't ever have ambiguity. The user will probably only ever see macro code in its pre-expanded or fully expanded states, and not any potentially confusing in-between states.

I don't think we have to worry about introspection because you can't introspect at all on these things until all declarations are created anyways.

Reply:
In the example here, I think that depends on whether we consider a library-level macro like `@generateXTopLevelVariable` to be a parent of the macros on top-level declarations. I think I was assuming here it would be, so that a library-level macro could see all of the declarations in the library.

Reply:
> I think I was assuming here it would be, so that a library-level macro could see all of the declarations in the library.

I think that the same rules still work out - since the introspective capabilities could still be limited to guarantee that. So for instance a Library macro would not be able to do subtype checks on types it sees, because other library type macros might still be messing with types.

Reply:
We can follow up and discuss this further